src/runtime-rs/crates/hypervisor/src/device/device_manager.rs

```diff
@@ -0,0 +1,338 @@
+    }
+
+    // declare the available block index
+    fn declare_device_index(&mut self) -> Result<u64> {
+        let current_index = if let Some(index) = self.released_block_index.pop() {
+            index
+        } else {
+            self.block_index
+        };
+        self.block_index += 1;
+
+        Ok(current_index)
+    }
+
+    fn release_device_index(&mut self, index: u64) {
+        self.released_block_index.push(index);
+        self.released_block_index.sort_by(|a, b| b.cmp(a));
+    }
+}
+
+#[derive(Clone)]
+pub struct DeviceManager {
+    pub dev_managers: HashMap<DeviceType, Arc<RwLock<Box<dyn DeviceManagerInner + Send + Sync>>>>,
+    hypervisor: Arc<dyn Hypervisor>,
+    shared_info: SharedInfo,
+}
+
+impl DeviceManager {
+    pub async fn new(hypervisor: Arc<dyn Hypervisor>) -> Result<Self> {
+        let mut managers =
+            HashMap::<DeviceType, Arc<RwLock<Box<dyn DeviceManagerInner + Send + Sync>>>>::new();
+        // register block device manager
+        let block_device_driver = hypervisor
+            .hypervisor_config()
+            .await
+            .blockdev_info
+            .block_device_driver;
+        managers.insert(
+            DeviceType::Block,
+            Arc::new(RwLock::new(Box::new(BlockDeviceManager::new(
+                &block_device_driver,
+            )?))),
+        );
+        // TODO: other device classes will be added later.
+        // https://github.com/kata-containers/kata-containers/issues/6525
+        // https://github.com/kata-containers/kata-containers/issues/6526
+        Ok(DeviceManager {
+            dev_managers: managers,
+            hypervisor,
+            shared_info: SharedInfo::new(1),
+        })
+    }
+
+    pub async fn try_add_device(
+        &mut self,
+        dev_info: &mut GenericConfig,
+        class: &DeviceType,
+    ) -> Result<String> {
+        let da = self
+            .make_device_argument(dev_info.dev_type.as_str())
+            .context("failed to make device arguments")?;
+        if let Some(device_manager) = self.dev_managers.get(class) {
+            match device_manager
+                .write()
+                .await
+                .try_add_device(dev_info, self.hypervisor.as_ref(), da.clone())
+                .await
+            {
+                Ok(device_id) => {
+                    return Ok(device_id);
+                }
+                Err(e) => {
+                    if let Some(index) = da.index {
+                        self.shared_info.release_device_index(index);
+                    }
+                    return Err(e);
+                }
+            };
+        }
+        if let Some(index) = da.index {
+            self.shared_info.release_device_index(index);
+        }
+        Err(anyhow!("invalid device class {:?}", class))
+    }
+
+    pub async fn try_remove_device(&mut self, device_id: String, class: &DeviceType) -> Result<()> {
+        if let Some(dev_manager) = self.dev_managers.get(class) {
```

Comment:
I am confused with the design of the `class: &DevicType`.  I think the `DeviceManager` should operate the device with the function of `trait Device`.

Reply:
`DeviceType` is to determine which kind of device manager we are going to use for the device